src/main/java/com/example/projectfirst/pipeline/PipelineService.java

```diff
@@ -11,46 +16,51 @@
 import java.util.Map;
 
 @Service
-public class PipelineService {
+public class PipelineService implements PipelineInterface{
     @Autowired
     private PipelineRepository pipelineRepository;
 
     public List<PipelineCollection> fetchAllPipelines() {
         return pipelineRepository.findAll();
     }
 
-    public String savePipeline(String yaml) {
+
+    public PipelineCollection fetchPipeline(String id) {
+        return pipelineRepository.findById(id)
+                .orElseThrow(() -> new PipelineNotFoundException(id));
+    }
+
+    public PipelineCollection savePipeline(String yaml) throws ObjectMapperException {
         // YAML to POJO
         ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
 
         try {
             Map<String, Pipeline> pipelineMap = objectMapper.readValue(yaml,
-                    new TypeReference<Map<String, Pipeline>>(){});
+                    new TypeReference<>(){});
             Pipeline pipe = pipelineMap.get("pipeline");
-            if(pipelineRepository.existsById(pipe.getId())){
-                return "Pipeline with that id already exists!";
+            String id = pipe.getId();
+
+            if(pipelineRepository.existsById(id)){
+                throw new PipelineAlreadyExistsException(id);
             }
-            PipelineCollection pipeline = new PipelineCollection(pipe.getId(), yaml, LocalDateTime.now(), LocalDateTime.now());
+            PipelineCollection pipeline = new PipelineCollection(id,yaml, LocalDateTime.now(), LocalDateTime.now());
             pipelineRepository.save(pipeline);
+            return pipeline;
         } catch (IOException e) {
-            e.printStackTrace();
-            return e.getMessage();
+            throw new ObjectMapperException();
```

Comment:
Add project abbreviation prefix to exception names, initialize exception with caught exception.
             throw new ObjectMapperException(); -> throw new ObjectMapperException(e);